tenacity/__init__.py

```diff
@@ -295,6 +305,10 @@ def iter(self, result, exc_info, start_time):  # noqa
         self.statistics['idle_for'] += sleep
         self.statistics['attempt_number'] += 1
 
+        if self.before_sleep is not None:
+            self.before_sleep(self.fn, self.statistics['attempt_number'],
```

Comment:
I wonder if it would not be simple to simple call `self.before_sleep(self, sleep=sleep, last_result=fut)` so you just have access to everything?

Reply:
Yea, that make the hook simpler and more general. I'll adapt the PR.